### PR TITLE
Source ZohoCRM: Update api version and default optional fields to None

### PR DESCRIPTION
**Pull request by richardlam-quotapath**
## What
The API Version was out of date which was causing issues because the v2 version has some modules return meta data with missing required fields.
Example: The deals endpoint with our customer's credentials was missing the "json_type" and "length" fields in the meta data which was causing the deals stream to not be available.
Also the optional fields in the FieldMeta class were not actually optional because they weren't being defaulted to None, local testing of that isolated code shows that if we have data missing the picklist_values field (which is supposed to be optional), the __init__() would fail because it wasn't being defaulted to None. 
image

## How
Fix #1
Update the zohocrm api version from v2 to v4
This will solve the ZohoCRM API returning data that is missing required fields issue

Fix #2
Have the class FieldMeta's optional fields be defaulted to None
This will solve the FieldMeta's __init__() failing on valid data due to missing optional fields issue

Re-testing shows the deals stream now available
image

## Recommended reading order
This shows that dataclasses optional fields must be defaulted to a value
https://stackoverflow.com/questions/70809438/python-dataclasses-with-optional-attributes

## 🚨 User Impact 🚨

Some of our clients were not seeing key streams such as Deals due to this issue

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests#pull-request-title-convention)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Review by sajarin — CHANGES_REQUESTED** [review]
Thanks for updating the api and cleaning up the code: 

Can you bump the version in the metadata.yaml file? 

Can you add a changelog entry to zohocrm.md in the docs directory?

**Comment by richardlam-quotapath** [comment]
> Thanks for updating the api and cleaning up the code:
> 
> Can you bump the version in the metadata.yaml file?
> 
> Can you add a changelog entry to zohocrm.md in the docs directory?

Yup bumped the version and added a changelog entry

**Comment by sajarin** [comment]
@richardlam-quotapath, apologies for the delay with merging this. Currently, this connector is not passing our integration tests due to some api regressions and other errors: https://connectors.airbyte.com/files/generated_reports/test_summary/source-zoho-crm/index.html

We need to fix these issues first before merging.

**Comment by richardlam-quotapath** [comment]
> @richardlam-quotapath, apologies for the delay with merging this. Currently, this connector is not passing our integration tests due to some api regressions and other errors: https://connectors.airbyte.com/files/generated_reports/test_summary/source-zoho-crm/index.html
> 
> We need to fix these issues first before merging.

Ah I see, it's not passing an acceptance test. I could take a look at the cause

File: airbyte-integrations/connectors/source-zoho-crm/source_zoho_crm/streams.py [hunk 1]
```diff
@@ -44,7 +44,8 @@ def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapp
         yield from data
 
     def path(self, *args, **kwargs) -> str:
-        return f"/crm/v2/{self.module.api_name}"
+        fields = ",".join([field.api_name for field in self.module.fields][0:50]) # Note, limited to 50 fields at max
+        return f"/crm/v4/{self.module.api_name}?fields={fields}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-zoho-crm/source_zoho_crm/streams.py:48]:
```suggestion
        if len(self.module) > 50:
            logger.warn("ZohoCRM only allow maximum of 50 parameter per query, the connector will drop: self.module.fields[50:]")
        fields = ",".join([field.api_name for field in self.module.fields][0:50]) 
        return f"/crm/v4/{self.module.api_name}?fields={fields}"
```

File: airbyte-integrations/connectors/source-zoho-crm/source_zoho_crm/streams.py [hunk 2]
```diff
@@ -92,16 +93,6 @@ def read_records(self, *args, **kwargs) -> Iterable[Mapping[str, Any]]:
...
 
-    def request_headers(
-        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
-    ) -> Mapping[str, Any]:
-        last_modified = stream_state.get(self.cursor_field, self._start_datetime)
-        # since API filters inclusively, we add 1 sec to prevent duplicate reads
-        last_modified_dt = datetime.datetime.fromisoformat(last_modified)
-        last_modified_dt += datetime.timedelta(seconds=1)
-        last_modified = last_modified_dt.isoformat("T", "seconds")
-        return {"If-Modified-Since": last_modified}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-zoho-crm/source_zoho_crm/streams.py:103]: Why are you removing this?

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Sorry the delay to return to you @richardlam-quotapath I left some comments. Let me know if there anything else I can do to help you. When you fix them we can merge your contribution.

**Comment by sajarin** [comment]
hey @richardlam-quotapath are you able to continue the PR? For now we'll close it and reopen once you have the time.

**Pull request closed without merging**